### ISSUE TITLE
Particle copySpecies is Slow

### ISSUE THREAD

**Issue by anshumang** [ComputationalRadiationPhysics/picongpu#858]
@ax3l @psychocoderHPC 
I was wondering if any of you had noticed my post on the picongpu-users mailing list. If not, here's the post (copied) -

I ran some measurements on the ADIOSWriter plugin and wanted to check if there are some reference numbers to validate against. I could only run it on a M2090 so numbers might not agree but still wanted to get a ballpark comparison.

Experiment description:
* -g 128 128 128
* -d 1 1 1
* Single node
* KelvinHelmholtzInstability

Performance data:
* Avg simulation timestep : 2.1sec
* ADIOSWriter : 338sec
    * Field : 1.8sec
    * Species1 : 165sec
        * kernel 'copySpecies' : 163.6sec
    * Species2 : 165.1sec
        * kernel 'copySpecies' : 163.6sec

Questions:
* Why is the destination buffer (deviceFrame) of 'copySpecies' alloc'd on host pinned memory and not on device memory?
* Does the 163sec of execution time of 'copySpecies' for the chosen simulation size look reasonable even for an M2090?
*  If the source buffer (speciesTmp->getDeviceParticlesBox()) is copied to host memory and a CPU version of 'copySpecies' is run instead, would it be same semantically?
-----> To do the above, I measured the following - 
           speciesTmp->synchronize();
           cudaDeviceSynchronize();
        It comes to -
             Species1 : 369 ms
             Species2 : 443 ms

**Comment by ax3l** [comment]
@anshumang yes we saw your mail in the [list](https://cg.hzdr.de/Lists/picongpu-users/Message/10.html) and it's absolutely the right place!
we just did not find the time to answer it yesterday evening, pls give us a second :)

**Comment by anshumang** [comment]
Great ! Will wait for your response.

**Comment by ax3l** [comment]
[Posted](https://cg.hzdr.de/Lists/picongpu-users/Message/11.html) :)

**Comment by anshumang** [comment]
Thanks a ton :+1: 
Getting back with the environment details shortly....

Key points, I think are (pls correct me where I am wrong) -
* Need to rerun on the K40
* Perform a deep copy of the species data otherwise the copy doesn't make sense

**Comment by anshumang** [comment]
Couldn't find a reply button on the mailing list viewer....so replying here with only the context ...

AH : the "copySpecies" kernel basically *is* a deep-copy.
AG : but to run it, the simulation has to be blocked, no?...have you considered, 
(1) deep copying all the data using the DMA engines from GPU memory to host memory 
(2) then run the strided access kernel on the CPU to populate a contiguous target buffer for adios_write
?

>   - what version of PIConGPU are you running
>     (beta-rc6, release-0.1.0, latest dev?)
Master, last commit : f0051b7 Hotfix: Race Condition in MW Solver & Pusher #604 on Jan 5, 2015
>   - what system/compilers/third-party libraries are you using?
OS : RHEL 6.6
g++ : 4.4.7
nvcc : 6.5
adios : 1.8.0

> 
> In the current development (dev) version of PIConGPU, we do not use
> double-buffering for particles any more. For that, we nowadays use
> *mapped* memory for access from the host side, e.g., for dumps.
Does this mean the host side can access parts of the species data before the copySpecies kernel completes?
 
> 
> How many time steps did you run?
10
> How often did you dump (adios.period)?
1
> How many particles did you use in total?
~100 macro particles/gpu, 1 GPU, 1 node (-d 1 1 1)
> 
> Also, the M2090 ("Fermi") generation is "pretty old" meaning it might be
> possible that the performance of the Fermi implementation for mapped
> memory is not comparable any more to modern hardware ("Kepler" is out
> since Nov/2012). Please use modern hardware for benchmarks, we won't
> optimize for that generation (even if we still support general operation
> on it).
> 
> We did not benchmark `copySpecies` in detail yet, but since we only use
> it on Kepler the "163.6sec" sound a bit unrealistic (or way off). Can
> you compare it to your Kepler cards in your node, pls?
> 
Will get back with this shortly.
> 
> The
>   speciesTmp->synchronize();
> is without implementation - it would usually synchronize data from
> device to host but is not implemented for particles (due to their nested
> memory structure, see above).
> 
> So basically what you measured is the time of a
>   cudaDeviceSynchronize();
> and *all the kernels that were still running* before that, which can be
> anything.
> 
> Also: before you measure kernels, always run a
>   cudaDeviceSynchronize();
> else you get the unpredictable load of asynchronous kernels that you
> don't intended to measure ;)
Hmmm...interesting...I do have a leading and trailing cudaDeviceSynchronize() like the following -
            cudaDeviceSynchronize();
            speciesTmp->synchronize();
            cudaDeviceSynchronize();
I am surprised that the duration for the above is ~400 ms even though the actual cudaMemcpy does not occur. Any ideas why this is happening?

**Comment by ax3l** [comment]
I contacted the listmaster about the problem posting on the list. for now let us continue here.

**Comment by ax3l** [comment]
**Software environment:**
When developing *new features* for PIConGPU, pls use the current `dev` branch to start from. We changed a lot since `beta-rc6` regarding the memory infrastructure, while introducing [mallocMC](https://github.com/ComputationalRadiationPhysics/mallocMC) and multiple-species support :)

gcc 4.4 (minor line release in 2009) is [pretty old for what CUDA 6.5 supports](https://gist.github.com/ax3l/9489132) and might not lead to optimal results. GCC 4.8 (2013) or 4.9 (2014) would generally be a better choic (requires CUDA und third-party lib recompiles).

be aware that all current *stable* releases of ADIOS, including 1.8.0, have problems with some compilers (such as gcc) when creating aggregated (local) files above 2GB: https://github.com/ornladios/ADIOS/issues/50 that might be fixed in the current `master` of ADIOS, I am currently evaluating that. Nevertheless, again the PIConGPU `dev` had a major refactoring and upgrade of our ADIOS output in #679 that you should base on.

> Aha...so I am going back to your older implementation :)
> What made you drop the double buffer approach? Wouldn't it allow the asynchronous copies to be > overlapped with the simulation timestep?

No please do not *go back*, you are already on an *old state*. Please advance to/start from the *latest* `dev`.

> `copySpecies`: but to run it, the simulation has to be blocked, no?...
> what was the problem with the earlier double buffer approach (as mentioned below) of -
> (1) deep copying all the data using the DMA engines from GPU memory to host memory knowing > that the DMA engines are free for most of the timestep
> (2) then run the strided access kernel on the CPU to populate a contiguous target buffer for
> adios_write?

-> @psychocoderHPC  for you

> AH: How many particles did you use in total?
> AG: ~100 macro particles/gpu, 1 GPU, 1 node (-d 1 1 1)

rly? only 100 (one hundred) in total? you are sure not something like 10e6?

> I am surprised that the duration for the above is ~400 ms even though the actual cudaMemcpy
> does not occur. Any ideas why this is happening?

As I said: you measured the wait-time for already running kernels with `cudaDeviceSynchronize()`. That is in general a rather random number.

or where did you put your measurement points in the chain of:
> A cudaDeviceSynchronize();
> B speciesTmp->synchronize();
> C cudaDeviceSynchronize();

B & C should not cause so long, but it might be possible that `cudaDeviceSynchronize()` is slow in general (that's why we don't pull it regularly).

> EDIT: Are you sure that the synchronize() is not implemented?

In general, it *is implemented* for all buffers ([FieldB](https://github.com/ComputationalRadiationPhysics/picongpu/blob/039819d2cf0f6b61fd2a33671dafeb6762d8de9a/src/picongpu/include/fields/FieldB.tpp#L118-L121), FieldE, FieldTmp, particles, ...) and all use `TaskCopyDeviceToHost`. But for particles, the implementation [is *empty*](https://github.com/ComputationalRadiationPhysics/picongpu/blob/039819d2cf0f6b61fd2a33671dafeb6762d8de9a/src/picongpu/include/particles/Particles.tpp#L119-L123).

What you see is the general event calls, but nothing happens for `speciesTmp` :)

**Comment by psychocoderHPC** [comment]
**What was the problem with the double buffer approach?**
**What made you drop the double buffer approach? 
Wouldn't it allow the asynchronous copies to be overlapped with the simulation timestep?**
We use a new memory allocator on the device which allows to use one shared memory buffer for particle species. Before every particle species had a fixed not shared memory buffer (this generate problems with multi species). 

**Why not copy species memory to host and run *strided access kernel on CPU*?**
The internel used native GPU pointer made it hard to iterate over the particle frames on the host side. 

**Are you sure that the synchronize() is not implemented?**
- in the old `beta-rc6` release `synchronize()` was implemented for particles
- in the current `dev` this method is empty for particle species.

**Comment by anshumang** [comment]
@ax3l  @psychocoderHPC 

Thanks a lot for the clarification as to why synchronize() was increasing the duration, I was not on *dev* :+1: 

AH : rly? only 100 (one hundred) in total? you are sure not something like 10e6?
AG : My bad...typo...100M macro particles per gpu.....so 100e6.....104857600, to be exact....

RW: The interally used native GPU pointer make it hard to iterate over the particle frames on the host side.
AG: From what I understand, the 'species size' is fixed for a given GPU based on available memory. The timestep duration increases with the 'grid size' but the time to transfer the species data would remain constant. The D2H bandwidth I measured on a K40 is ~1.67GB/s making it possible to transfer the species data inside one timestep for a large enough 'grid size'. Can this be a motivation to modify the particle data structure so that it can be iterated on the host? This would avoid running copySpecies on the device that blocks simulation while it runs.

**Comment by psychocoderHPC** [comment]
With ~1.67Gbyte/s and up to 12GByte memory on a K40 we need ~6 sec to copy the particle data to the host memory.
In the current dev we use one big memory blob for all particle species. First all fields are allocated and the rest of the memory is allocated for the species.
The memory for one species is not contiguous, therefor you must copy the full species memory blob to the device. This can't run in parallel to the simulation due to the fact that the simulation permanently write to the memory blob.

> From what I understand, the 'species size' is fixed for a given GPU based on available memory. 

The number of species is fixed but the number of particles per species vary because particles are moving.

**Comment by psychocoderHPC** [comment]
>  Can this be a motivation to modify the particle data structure so that it can be iterated on the host?

I opened a issue for this feature request #864

**Comment by anshumang** [comment]
:+1: 
I will work on this.

**Comment by ax3l** [comment]
> AH : rly? only 100 (one hundred) in total? you are sure not something like 10e6?
> AG : My bad...typo...100M macro particles per gpu.....so 100e6.....104857600, to be exact....

a good measure is to use not more than ~10e6 particles per GB of GPU ram in the old implementation that you were using.

just to make sure you are not overfloating the GPU RAM, which would obfuscate all the results with undefined behavior, try to reduce the number of particles a little.

**Comment by anshumang** [comment]
I was relying on the error message 'Ringbuffer: memory overflow'. I chose the problem size one half  of the max size which did not throw the error message.
-g 256 256 128 => 400e6 => Error
-g 256 128 128 => 200e6 => No error
-g 128 128 128 => 100e6 => No error (chosen size)

Is the ~10e6 still an upper limit for the latest *dev*?

**Comment by psychocoderHPC** [comment]
No 10e6 is not the upper limit. It is you save number for the most cases. Keep in mind this is 10e6 per Gbyte of device memory.

The current dev can handle more particles than the old ringbuffer.

The Ringbuffer error messages in the master und upcomming release were written with printf from the kernel. Do to the fact that we not synchronize the device while we run one timestep the error is not always shown.

**Comment by anshumang** [comment]
I reran the experiment with ~13e6 macro particles per gpu on a single node with a K40 using release-0.1.0.

Without adios => avg time step ~200 ms
./build_picongpu/picongpu -g 64 64 64 -d 1 1 1 -s 10
PIConGPUVerbose PHYSICS(1) | Sliding Window is OFF
freeGpuMem : 11838676992
mem for particles=4095 MiB = 540383 Frames = 138338048 Particles
mem for particles=4095 MiB = 540383 Frames = 138338048 Particles
PIConGPUVerbose PHYSICS(1) | max weighting 326.577
PIConGPUVerbose PHYSICS(1) | Courant c*dt <= 1.19997 ? 1
PIConGPUVerbose PHYSICS(1) | omega_pe * dt <= 0.1 ? 0.0267598
PIConGPUVerbose PHYSICS(1) | macro particles per gpu: 13107200
PIConGPUVerbose PHYSICS(1) | typical macro particle weighting: 326.577
PIConGPUVerbose PHYSICS(1) | UNIT_SPEED 2.99792e+08
PIConGPUVerbose PHYSICS(1) | UNIT_TIME 1.5e-16
PIConGPUVerbose PHYSICS(1) | UNIT_LENGTH 4.49689e-08
PIConGPUVerbose PHYSICS(1) | UNIT_MASS 2.97492e-28
PIConGPUVerbose PHYSICS(1) | UNIT_CHARGE 5.23234e-17
PIConGPUVerbose PHYSICS(1) | UNIT_EFIELD 1.13634e+13
PIConGPUVerbose PHYSICS(1) | UNIT_BFIELD 37904.2
PIConGPUVerbose PHYSICS(1) | UNIT_ENERGY 2.67372e-11
initialization time:  1min  8sec 384msec = 68 sec
  0 % =        0 | time elapsed:                    0msec | avg time per step:   0msec
 10 % =        1 | time elapsed:                  168msec | avg time per step: 168msec
 20 % =        2 | time elapsed:                  354msec | avg time per step: 185msec
 30 % =        3 | time elapsed:                  526msec | avg time per step: 171msec
 40 % =        4 | time elapsed:                  714msec | avg time per step: 187msec
 50 % =        5 | time elapsed:                  888msec | avg time per step: 173msec
 60 % =        6 | time elapsed:             1sec  74msec | avg time per step: 185msec
 70 % =        7 | time elapsed:             1sec 249msec | avg time per step: 175msec
 80 % =        8 | time elapsed:             1sec 439msec | avg time per step: 190msec
 90 % =        9 | time elapsed:             1sec 615msec | avg time per step: 175msec
100 % =       10 | time elapsed:             1sec 811msec | avg time per step: 196msec
calculation  simulation time:  1sec 811msec = 1 sec
full simulation time:  1min 12sec 223msec = 72 sec


With adios => avg time step ~40 sec
./build_picongpu/picongpu -g 64 64 64 -d 1 1 1 -s 10 --adios.period 1
PIConGPUVerbose PHYSICS(1) | Sliding Window is OFF
freeGpuMem : 11838676992
mem for particles=4095 MiB = 540383 Frames = 138338048 Particles
mem for particles=4095 MiB = 540383 Frames = 138338048 Particles
PIConGPUVerbose PHYSICS(1) | max weighting 326.577
PIConGPUVerbose PHYSICS(1) | Courant c*dt <= 1.19997 ? 1
PIConGPUVerbose PHYSICS(1) | omega_pe * dt <= 0.1 ? 0.0267598
PIConGPUVerbose PHYSICS(1) | macro particles per gpu: 13107200
PIConGPUVerbose PHYSICS(1) | typical macro particle weighting: 326.577
PIConGPUVerbose PHYSICS(1) | UNIT_SPEED 2.99792e+08
PIConGPUVerbose PHYSICS(1) | UNIT_TIME 1.5e-16
PIConGPUVerbose PHYSICS(1) | UNIT_LENGTH 4.49689e-08
PIConGPUVerbose PHYSICS(1) | UNIT_MASS 2.97492e-28
PIConGPUVerbose PHYSICS(1) | UNIT_CHARGE 5.23234e-17
PIConGPUVerbose PHYSICS(1) | UNIT_EFIELD 1.13634e+13
PIConGPUVerbose PHYSICS(1) | UNIT_BFIELD 37904.2
PIConGPUVerbose PHYSICS(1) | UNIT_ENERGY 2.67372e-11
initialization time: 59sec 122msec = 59 sec
  0 % =        0 | time elapsed:            38sec 767msec | avg time per step:   0msec
 10 % =        1 | time elapsed:            38sec 970msec | avg time per step: 203msec
 20 % =        2 | time elapsed:       1min 17sec 689msec | avg time per step: 185msec
 30 % =        3 | time elapsed:       1min 56sec 293msec | avg time per step: 172msec
 40 % =        4 | time elapsed:       2min 34sec 906msec | avg time per step: 187msec
 50 % =        5 | time elapsed:       3min 13sec 408msec | avg time per step: 174msec
 60 % =        6 | time elapsed:       3min 51sec 895msec | avg time per step: 186msec
 70 % =        7 | time elapsed:       4min 30sec 236msec | avg time per step: 175msec
 80 % =        8 | time elapsed:       5min  8sec 570msec | avg time per step: 190msec
 90 % =        9 | time elapsed:       5min 46sec 819msec | avg time per step: 176msec
100 % =       10 | time elapsed:       6min 25sec  74msec | avg time per step: 196msec
calculation  simulation time:  7min  3sec  27msec = 423 sec
full simulation time:  8min  3sec 832msec = 483 sec

Another point -
Looks like there is an overflow occurring in the computation of 'avg time per step'

**Comment by psychocoderHPC** [comment]
Avg time per step shows only the time for the simulation without plugins.

What kind of storage do you use? Do you know how fast your storage is?
46sec per dump contains deep copy of all species over mapped memory to the host and write the data to the storage via adios.

**Comment by anshumang** [comment]
Ah got the point about 'avg time per step'...!

I set transport to NULL in adios_select_method and no .bp files are generated, so there shouldn't be any disk write, no?

**Comment by anshumang** [comment]
With adios (MPI_AGGREGATE) => avg. time step ~46-48 ms
./build_picongpu/picongpu -g 64 64 64 -d 1 1 1 -s 10 --adios.period 1
PIConGPUVerbose PHYSICS(1) | Sliding Window is OFF
mem for particles=4095 MiB = 540383 Frames = 138338048 Particles
mem for particles=4095 MiB = 540383 Frames = 138338048 Particles
PIConGPUVerbose PHYSICS(1) | max weighting 326.577
PIConGPUVerbose PHYSICS(1) | Courant c*dt <= 1.19997 ? 1
PIConGPUVerbose PHYSICS(1) | omega_pe * dt <= 0.1 ? 0.0267598
PIConGPUVerbose PHYSICS(1) | macro particles per gpu: 13107200
PIConGPUVerbose PHYSICS(1) | typical macro particle weighting: 326.577
PIConGPUVerbose PHYSICS(1) | UNIT_SPEED 2.99792e+08
PIConGPUVerbose PHYSICS(1) | UNIT_TIME 1.5e-16
PIConGPUVerbose PHYSICS(1) | UNIT_LENGTH 4.49689e-08
PIConGPUVerbose PHYSICS(1) | UNIT_MASS 2.97492e-28
PIConGPUVerbose PHYSICS(1) | UNIT_CHARGE 5.23234e-17
PIConGPUVerbose PHYSICS(1) | UNIT_EFIELD 1.13634e+13
PIConGPUVerbose PHYSICS(1) | UNIT_BFIELD 37904.2
PIConGPUVerbose PHYSICS(1) | UNIT_ENERGY 2.67372e-11
initialization time:  1min 11sec 981msec = 71 sec
  0 % =        0 | time elapsed:            42sec 968msec | avg time per step:   0msec
 10 % =        1 | time elapsed:            43sec 828msec | avg time per step: 859msec
 20 % =        2 | time elapsed:       2min 10sec 954msec | avg time per step: 186msec
 30 % =        3 | time elapsed:       3min  4sec 899msec | avg time per step:  1sec  60msec
 40 % =        4 | time elapsed:       3min 50sec 260msec | avg time per step: 819msec
 50 % =        5 | time elapsed:       4min 40sec 330msec | avg time per step: 659msec
 60 % =        6 | time elapsed:       5min 27sec 694msec | avg time per step: 209msec
 70 % =        7 | time elapsed:       6min 13sec 896msec | avg time per step: 202msec
 80 % =        8 | time elapsed:       7min  0sec 378msec | avg time per step: 239msec
 90 % =        9 | time elapsed:       7min 54sec 727msec | avg time per step: 240msec
100 % =       10 | time elapsed:       8min 43sec 945msec | avg time per step: 620msec
calculation  simulation time:  9min 35sec 180msec = 575 sec
full simulation time: 10min 50sec 463msec = 650 sec

**Comment by anshumang** [comment]
There's some variability when disk I/O is enabled so I timed the copySpecies as follows -
cudaDeviceSynchronize()
gettimeofday(&start, NULL)
copySpecies<<<>>>
cudaDeviceSynchronize()
gettimeofday(&end, NULL)

It takes ~ 25-30 ms to execute the copySpecies kernel. 

./build_picongpu/picongpu -g 64 64 64 -d 1 1 1 -s 10 --adios.period 1
PIConGPUVerbose PHYSICS(1) | Sliding Window is OFF
mem for particles=4095 MiB = 540383 Frames = 138338048 Particles
mem for particles=4095 MiB = 540383 Frames = 138338048 Particles
PIConGPUVerbose PHYSICS(1) | max weighting 326.577
PIConGPUVerbose PHYSICS(1) | Courant c*dt <= 1.19997 ? 1
PIConGPUVerbose PHYSICS(1) | omega_pe * dt <= 0.1 ? 0.0267598
PIConGPUVerbose PHYSICS(1) | macro particles per gpu: 13107200
PIConGPUVerbose PHYSICS(1) | typical macro particle weighting: 326.577
PIConGPUVerbose PHYSICS(1) | UNIT_SPEED 2.99792e+08
PIConGPUVerbose PHYSICS(1) | UNIT_TIME 1.5e-16
PIConGPUVerbose PHYSICS(1) | UNIT_LENGTH 4.49689e-08
PIConGPUVerbose PHYSICS(1) | UNIT_MASS 2.97492e-28
PIConGPUVerbose PHYSICS(1) | UNIT_CHARGE 5.23234e-17
PIConGPUVerbose PHYSICS(1) | UNIT_EFIELD 1.13634e+13
PIConGPUVerbose PHYSICS(1) | UNIT_BFIELD 37904.2
PIConGPUVerbose PHYSICS(1) | UNIT_ENERGY 2.67372e-11
initialization time:  1min 31sec 560msec = 91 sec
copySpecies(us) : 12601987
copySpecies(us) : 12607124
  0 % =        0 | time elapsed:            33sec 136msec | avg time per step:   0msec
 10 % =        1 | time elapsed:            34sec 210msec | avg time per step:  1sec  40msec
copySpecies(us) : 12609378
copySpecies(us) : 12608493
 20 % =        2 | time elapsed:       1min 15sec 657msec | avg time per step: 789msec
copySpecies(us) : 12561310
copySpecies(us) : 12563507
 30 % =        3 | time elapsed:       1min 48sec 888msec | avg time per step: 194msec
copySpecies(us) : 12565713
copySpecies(us) : 12571453
 40 % =        4 | time elapsed:       2min 29sec 400msec | avg time per step: 187msec
copySpecies(us) : 12546891
copySpecies(us) : 12525664
 50 % =        5 | time elapsed:       3min  9sec 177msec | avg time per step: 190msec
copySpecies(us) : 14304455
copySpecies(us) : 14677379
 60 % =        6 | time elapsed:       3min 49sec 468msec | avg time per step: 186msec
copySpecies(us) : 14718796
copySpecies(us) : 14726960
 70 % =        7 | time elapsed:       4min 28sec 747msec | avg time per step: 202msec
copySpecies(us) : 12490348
copySpecies(us) : 12477308
 80 % =        8 | time elapsed:       5min 13sec 816msec | avg time per step: 191msec
copySpecies(us) : 14989935
copySpecies(us) : 12710819
 90 % =        9 | time elapsed:       5min 50sec 883msec | avg time per step: 189msec
copySpecies(us) : 14916553
copySpecies(us) : 12451161
100 % =       10 | time elapsed:       6min 25sec 665msec | avg time per step: 197msec
copySpecies(us) : 13097578
copySpecies(us) : 13633738
calculation  simulation time:  6min 59sec 397msec = 419 sec
full simulation time:  8min 33sec 969msec = 513 sec

**Comment by psychocoderHPC** [comment]
I think you measured 20 to 30 **sec** not **ms**

**Comment by ax3l** [comment]
when you guys edit messages, it helps to use the markup:
```
~~strike through~~ that is new
```

which creates something like:

~~strike through~~ that is new

**Comment by anshumang** [comment]
@ax3l will do from here on...

**Comment by ax3l** [comment]
@anshumang sure no problem, just a hint so we don't get confused :)

**Comment by psychocoderHPC** [comment]
Here are the results of a quick test on our cluster at HZDR:
- ~12million particles per Species (~397 MiB)
- 2 species
- 1 GPU

DataHeader:`"Time(%)","Time","Calls","Avg","Min","Max","Name,%,s,,s,s,s,`

```BASH
widera@kepler003:/bigdata/hplsim/scratch/widera/build/params/bin$ nvprof -u s --csv -s ./picongpu -g 128 64 64 -d 1 1 1 -s 1 --periodic 1 1 1 --hdf5.period 1 2>&1 | grep copySpecies
47.014586,15.141966,2,7.570983,7.527426,7.614539,"void picongpu::copySpecies<PMacc::Frame<picongpu::OperatorCreateVectorBox, PMacc::ParticleDescription<boost::mpl::string<int=105, int=0, int=0, int=0, int=0, int=0, int=0, int=0>, PMacc::math::CT::Vector<mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=4>>, boost::mpl::v_item<picongpu::placeholder_definition17::globalCellIdx<picongpu::placeholder_definition24::globalCellIdx_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::v_item<picongpu::placeholder_definition21::weighting, boost::mpl::v_item<picongpu::placeholder_definition19::momentum, boost::mpl::v_item<picongpu::placeholder_definition16::position<picongpu::placeholder_definition18::position_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::vector0<mpl_::na>, int=0>, int=0>, int=0>, int=0>, boost::mpl::vector<picongpu::placeholder_definition26::particlePusher<picongpu::particles::pusher::Boris, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::placeholder_definition25::shape<picongpu::particles::pusher::shapes::TSC, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::placeholder_definition28::interpolation<picongpu::FieldToParticleInterpolation<picongpu::placeholder_definition25::shape<picongpu::particles::pusher::shapes::TSC, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::AssignedTrilinearInterpolation>, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::placeholder_definition29::current<picongpu::currentSolver::Esirkepov<picongpu::placeholder_definition25::shape<picongpu::particles::pusher::shapes::TSC, PMacc::placeholder_definition15::pmacc_isAlias>, unsigned int=3>, PMacc::placeholder_definition15::pmacc_isAlias>, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, boost::mpl::v_item<picongpu::CommunicationId<unsigned int=4>, boost::mpl::v_item<picongpu::placeholder_definition16::position<picongpu::placeholder_definition18::position_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::vector0<mpl_::na>, int=0>, int=0>>>, PMacc::ParticlesBox<PMacc::Frame<PMacc::ParticlesBuffer<PMacc::ParticleDescription<boost::mpl::string<int=105, int=0, int=0, int=0, int=0, int=0, int=0, int=0>, PMacc::math::CT::Vector<mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=4>>, boost::mpl::v_item<picongpu::placeholder_definition17::globalCellIdx<picongpu::placeholder_definition24::globalCellIdx_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::v_item<picongpu::placeholder_definition21::weighting, boost::mpl::v_item<picongpu::placeholder_definition19::momentum, boost::mpl::v_item<picongpu::placeholder_definition16::position<picongpu::placeholder_definition18::position_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::vector0<mpl_::na>, int=0>, int=0>, int=0>, int=0>, picongpu::CommunicationId, PMacc::ParticleDescription<boost::mpl::string<int=105, int=0, int=0, int=0, int=0, int=0, int=0, int=0>, PMacc::math::CT::Vector<mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=4>>, boost::mpl::v_item<picongpu::placeholder_definition17::globalCellIdx<picongpu::placeholder_definition24::globalCellIdx_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::v_item<picongpu::placeholder_definition21::weighting, boost::mpl::v_item<picongpu::placeholder_definition19::momentum, boost::mpl::v_item<picongpu::placeholder_definition16::position<picongpu::placeholder_definition18::position_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::vector0<mpl_::na>, int=0>, int=0>, int=0>, int=0>, boost::mpl::vector<picongpu::placeholder_definition26::particlePusher<picongpu::particles::pusher::Boris, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::placeholder_definition25::shape<picongpu::particles::pusher::shapes::TSC, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::placeholder_definition28::interpolation<picongpu::FieldToParticleInterpola"
42.216306,13.596586,2,6.798293,5.976638,7.619948,"void picongpu::copySpecies<PMacc::Frame<picongpu::OperatorCreateVectorBox, PMacc::ParticleDescription<boost::mpl::string<int=101, int=0, int=0, int=0, int=0, int=0, int=0, int=0>, PMacc::math::CT::Vector<mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=4>>, boost::mpl::v_item<picongpu::placeholder_definition17::globalCellIdx<picongpu::placeholder_definition24::globalCellIdx_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::v_item<picongpu::placeholder_definition21::weighting, boost::mpl::v_item<picongpu::placeholder_definition19::momentum, boost::mpl::v_item<picongpu::placeholder_definition16::position<picongpu::placeholder_definition18::position_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::vector0<mpl_::na>, int=0>, int=0>, int=0>, int=0>, boost::mpl::vector<picongpu::placeholder_definition26::particlePusher<picongpu::particles::pusher::Boris, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::placeholder_definition25::shape<picongpu::particles::pusher::shapes::TSC, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::placeholder_definition28::interpolation<picongpu::FieldToParticleInterpolation<picongpu::placeholder_definition25::shape<picongpu::particles::pusher::shapes::TSC, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::AssignedTrilinearInterpolation>, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::placeholder_definition29::current<picongpu::currentSolver::Esirkepov<picongpu::placeholder_definition25::shape<picongpu::particles::pusher::shapes::TSC, PMacc::placeholder_definition15::pmacc_isAlias>, unsigned int=3>, PMacc::placeholder_definition15::pmacc_isAlias>, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, boost::mpl::v_item<picongpu::CommunicationId<unsigned int=5>, boost::mpl::v_item<picongpu::placeholder_definition16::position<picongpu::placeholder_definition18::position_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::vector0<mpl_::na>, int=0>, int=0>>>, PMacc::ParticlesBox<PMacc::Frame<PMacc::ParticlesBuffer<PMacc::ParticleDescription<boost::mpl::string<int=101, int=0, int=0, int=0, int=0, int=0, int=0, int=0>, PMacc::math::CT::Vector<mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=4>>, boost::mpl::v_item<picongpu::placeholder_definition17::globalCellIdx<picongpu::placeholder_definition24::globalCellIdx_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::v_item<picongpu::placeholder_definition21::weighting, boost::mpl::v_item<picongpu::placeholder_definition19::momentum, boost::mpl::v_item<picongpu::placeholder_definition16::position<picongpu::placeholder_definition18::position_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::vector0<mpl_::na>, int=0>, int=0>, int=0>, int=0>, picongpu::CommunicationId, PMacc::ParticleDescription<boost::mpl::string<int=101, int=0, int=0, int=0, int=0, int=0, int=0, int=0>, PMacc::math::CT::Vector<mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=8>, mpl_::integral_c<int, int=4>>, boost::mpl::v_item<picongpu::placeholder_definition17::globalCellIdx<picongpu::placeholder_definition24::globalCellIdx_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::v_item<picongpu::placeholder_definition21::weighting, boost::mpl::v_item<picongpu::placeholder_definition19::momentum, boost::mpl::v_item<picongpu::placeholder_definition16::position<picongpu::placeholder_definition18::position_pic, PMacc::placeholder_definition15::pmacc_isAlias>, boost::mpl::vector0<mpl_::na>, int=0>, int=0>, int=0>, int=0>, boost::mpl::vector<picongpu::placeholder_definition26::particlePusher<picongpu::particles::pusher::Boris, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::placeholder_definition25::shape<picongpu::particles::pusher::shapes::TSC, PMacc::placeholder_definition15::pmacc_isAlias>, picongpu::placeholder_definition28::interpolation<picongpu::FieldToParticleInterpola"
```

**I can confirm that `copySpecies` is very slow**

I need 7,5sec to copy 397MiB results in ~53  MiB/s bandwith.

**Comment by ax3l** [comment]
Related issues:
- `mallocMC` implementation: https://github.com/ComputationalRadiationPhysics/mallocMC/pull/86 #906
- libPMacc kernel mapper host-side support: #902
- host-side particle access #905
- host-side OpenMP: #908

**Comment by psychocoderHPC** [comment]
solved with #907

**Comment by ax3l** [comment]
@anshumang if you are still hanging in: the problem that you spotted should now be resolved! :+1: 

feel free to open a new issue for follow-up questions if they arise!

**Comment by anshumang** [comment]
:+1:  great work, @ax3l @psychocoderHPC Can you explain some more about the fix?

**Comment by psychocoderHPC** [comment]
Some more information are coming next week

**Comment by anshumang** [comment]
Thanks for the reply @psychocoderHPC

**Comment by ax3l** [comment]
@psychocoderHPC pls :)

**Comment by psychocoderHPC** [comment]
Sry for the late response. Now the first information are coming.

**MallocMCBuffer:**
- provide double buffer of mallocMC heap on host side
- handle synchronization
- store offset of virtual host/device address

**ParticlesBox:**
- handle creation, destruct and iteration over frames
- *device side:*
  - creation of frames
  - destruct of frames
  - iteration over frames
- *host side:* (read only, because we don't know the heap structure)
  - iteration over frames
  - pointer of linked list are translated on the fly

**Frame:**
- contains all attributes of a species
- max number of particles per frame is equal to the superCell size
- contains plain C pointer to
  - next frame
  - previous frame

**SuperCell:**
- contains first pointer and last pointer to frame list
- meta information like `sizeLastFrame` (number of particle in the last frame)

**ParticlesBuffer:**
- contains grid of `SuperCell`'s (entry and exit points of linked list)
- create `ParticlesBox`


**How to read traverse frame lists on the host side:**
- copy mallocMC heap to the host with `MallocMCBuffer::synchronize()`
  - `synchronize()` is called wire our `DataConnector` https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/picongpu/include/plugins/adios/ADIOSWriter.hpp#L675
  - the heap contains all frame of each species
  - there is no knowledge present were a frame starts or end (this is defined by the meta data of a superCell)
  - the host side double buffer of the heap from mallocMC is created on the first call of `synchronize()`
- copy SuperCell meta information to the host
  - `synchronize()` must called for each species https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/picongpu/include/plugins/adios/ADIOSWriter.hpp#L681-L682
  - each species has it's own superCell meta data!
- for adios output:
  - the host side `ParticleBox` needs the memory offset (int bytes) to map device pointer to host pointer inside the `MallocMCBuffer` https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/picongpu/include/plugins/adios/WriteSpecies.hpp#L139
  - iterate over frame list and concate list of frames https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/picongpu/include/plugins/kernel/CopySpecies.kernel#L66
  - the method how to iterate over frame lists look nearly equal to the [device side code](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/picongpu/include/plugins/kernel/CopySpecies.kernel#L155)

**Notes:**
- the user (PIConGPU) must manually call 
  - `MallocMCBuffer::synchronize()` 
  - `synchronize()` for each species
  - if both methods are called in different time steps the meta data per species and MallocMCBuffer not match (unknown behaviour)
- currently only the adios plugin uses the `MallocMCBuffer`
  - if more plugins will use the `MallocMCBuffer` the buffer must only synchronized one time per step

**Comment by psychocoderHPC** [comment]
Here is a small figure out of my lab book which shows the address mapping of the pointers:

![mallocmchostbuffer](https://cloud.githubusercontent.com/assets/4037125/9329782/d8aaa154-45b4-11e5-9518-6c801ae885c9.png)

**Comment by anshumang** [comment]
@psychocoderHPC :+1:  thanks for sharing all the details...I am trying to parse them now....

**Comment by anshumang** [comment]
first measurement on single node K40 .... adds 3-4 sec per timestep if I comment out the beginAdios, writeAdios and endAdios calls...
./picongpu -d 1 1 1 -g 64 64 64 -s 10 --adios.period 1
PIConGPUVerbose PHYSICS(1) | Sliding Window is OFF
PIConGPUVerbose PHYSICS(1) | Courant c*dt <= 1.00556 ? 1
PIConGPUVerbose PHYSICS(1) | species e: omega_p * dt <= 0.1 ? 0.0319333
PIConGPUVerbose PHYSICS(1) | species i: omega_p * dt <= 0.1 ? 0.000745229
PIConGPUVerbose PHYSICS(1) | macro particles per gpu: 13107200
PIConGPUVerbose PHYSICS(1) | typical macro particle weighting: 326.577
PIConGPUVerbose PHYSICS(1) | UNIT_SPEED 2.99792e+08
PIConGPUVerbose PHYSICS(1) | UNIT_TIME 1.79e-16
PIConGPUVerbose PHYSICS(1) | UNIT_LENGTH 5.36628e-08
PIConGPUVerbose PHYSICS(1) | UNIT_MASS 2.97492e-28
PIConGPUVerbose PHYSICS(1) | UNIT_CHARGE 5.23234e-17
PIConGPUVerbose PHYSICS(1) | UNIT_EFIELD 9.5224e+12
PIConGPUVerbose PHYSICS(1) | UNIT_BFIELD 31763.3
PIConGPUVerbose PHYSICS(1) | UNIT_ENERGY 2.67372e-11
initialization time: 10sec 452msec = 10 sec
  0 % =        0 | time elapsed:            14sec 521msec | avg time per step:   0msec
 10 % =        1 | time elapsed:            14sec 701msec | avg time per step: 179msec
 20 % =        2 | time elapsed:            17sec 567msec | avg time per step: 174msec
 30 % =        3 | time elapsed:            20sec 439msec | avg time per step: 181msec
 40 % =        4 | time elapsed:            23sec 305msec | avg time per step: 175msec
 50 % =        5 | time elapsed:            26sec 178msec | avg time per step: 181msec
 60 % =        6 | time elapsed:            29sec  51msec | avg time per step: 182msec
 70 % =        7 | time elapsed:            31sec 929msec | avg time per step: 186msec
 80 % =        8 | time elapsed:            34sec 808msec | avg time per step: 187msec
 90 % =        9 | time elapsed:            37sec 677msec | avg time per step: 178msec
100 % =       10 | time elapsed:            40sec 554msec | avg time per step: 185msec
calculation  simulation time: 43sec 245msec = 43 sec
full simulation time: 54sec 294msec = 54 sec

Leaving the *Adios calls, causes an insufficient memory crash for anything more then -g 64 32 32 (3.3 particles per GPU) -
./picongpu -d 1 1 1 -g 64 64 32 -s 10 --adios.period 1
PIConGPUVerbose PHYSICS(1) | Sliding Window is OFF
PIConGPUVerbose PHYSICS(1) | Courant c*dt <= 1.00556 ? 1
PIConGPUVerbose PHYSICS(1) | species e: omega_p * dt <= 0.1 ? 0.0319333
PIConGPUVerbose PHYSICS(1) | species i: omega_p * dt <= 0.1 ? 0.000745229
PIConGPUVerbose PHYSICS(1) | macro particles per gpu: 6553600
PIConGPUVerbose PHYSICS(1) | typical macro particle weighting: 326.577
PIConGPUVerbose PHYSICS(1) | UNIT_SPEED 2.99792e+08
PIConGPUVerbose PHYSICS(1) | UNIT_TIME 1.79e-16
PIConGPUVerbose PHYSICS(1) | UNIT_LENGTH 5.36628e-08
PIConGPUVerbose PHYSICS(1) | UNIT_MASS 2.97492e-28
PIConGPUVerbose PHYSICS(1) | UNIT_CHARGE 5.23234e-17
PIConGPUVerbose PHYSICS(1) | UNIT_EFIELD 9.5224e+12
PIConGPUVerbose PHYSICS(1) | UNIT_BFIELD 31763.3
PIConGPUVerbose PHYSICS(1) | UNIT_ENERGY 2.67372e-11
initialization time:  9sec 996msec = 9 sec
7365737
ERROR: adios_allocate_buffer (): insufficient memory: 295698432 requested, 282054656 available.  Using available.
terminate called after throwing an instance of 'std::runtime_error'
  what():  ADIOS: error at cmd 'adios_allocate_buffer(ADIOS_BUFFER_ALLOC_NOW,buffer_mem)' (-1, -1) in /opt/data1/share/users/anshuman/ComputationalRadiationPhysics/picongpu-repos/08-31-2015/picongpu/src/p
icongpu/include/plugins/adios/ADIOSWriter.hpp:981 adios_allocate_buffer (): insufficient memory: 295698432 requested, 282054656 available.  Using available.

[shiva:03637] *** Process received signal ***
[shiva:03637] Signal: Aborted (6)
[shiva:03637] Signal code:  (-6)
[shiva:03637] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x10340) [0x7f7d1f505340]
[shiva:03637] [ 1] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x39) [0x7f7d1d982cc9]
[shiva:03637] [ 2] /lib/x86_64-linux-gnu/libc.so.6(abort+0x148) [0x7f7d1d9860d8]
[shiva:03637] [ 3] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(_ZN9__gnu_cxx27__verbose_terminate_handlerEv+0x155) [0x7f7d1e1966b5]
[shiva:03637] [ 4] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x5e836) [0x7f7d1e194836]
[shiva:03637] [ 5] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x5e863) [0x7f7d1e194863]
[shiva:03637] [ 6] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x5eaa2) [0x7f7d1e194aa2]
[shiva:03637] [ 7] ./picongpu(_ZN8picongpu5adios11ADIOSWriter10writeAdiosEPvSs+0x2350) [0x891d9c]
[shiva:03637] [ 8] ./picongpu(_ZN8picongpu5adios11ADIOSWriter20notificationReceivedEjb+0x847) [0x88ac43] 
[shiva:03637] [ 9] ./picongpu(_ZN8picongpu5adios11ADIOSWriter6notifyEj+0x25) [0x887cc7]
[shiva:03637] [10] ./picongpu(_ZN5PMacc15PluginConnector13notifyPluginsEj+0x77) [0x8731dd]
[shiva:03637] [11] ./picongpu(_ZN5PMacc16SimulationHelperILj3EE11dumpOneStepEj+0x41) [0x8eb313]
[shiva:03637] [12] ./picongpu(_ZN5PMacc16SimulationHelperILj3EE15startSimulationEv+0x138) [0x8d43a6]
[shiva:03637] [13] ./picongpu(_ZN8picongpu17SimulationStarterINS_21InitialiserControllerENS_16PluginControllerENS_12MySimulationEE5startEv+0xad) [0x8bc31d]
[shiva:03637] [14] ./picongpu(main+0x140) [0x85940f]
[shiva:03637] [15] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7f7d1d96dec5]
[shiva:03637] [16] ./picongpu() [0x836fa7]
[shiva:03637] *** End of error message ***
Aborted (core dumped)
Is this known?

**Comment by ax3l** [comment]
@anshumang this is simply caused by having not enough memory on the host ram, a ratio 2:1 is at least recommended for host ram to sum of all GPU's RAM on a node.

**Comment by anshumang** [comment]
@ax3l Got the point for the insufficient memory issue....any benchmarking reference for the no-Adios copy-to-host timing?

**Comment by ax3l** [comment]
yes, @psychocoderHPC added some measurements in #908

**Comment by anshumang** [comment]
@psychocoderHPC 
I get 2,667 s to for this 
https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/picongpu/include/plugins/adios/ADIOSWriter.hpp#L673

In terms of #908 
# once
create malloc  2sec  49msec = 2.04942 sec
# per dump
copy mallocMC 713msec = 0.713393 sec
copy supercell e -  0msec = 4.49219e-05 sec
copy supercell i -  0msec = 5.29785e-05 sec
host copy kernel e - 645msec = 0.64594 sec
host copy kernel i - 655msec = 0.65568 sec
which one is it?
